### PR TITLE
[test] add traffic analysis for Cert_5_2_3_LeaderReject2Hops

### PR DESCRIPTION
**Pull request by hjian2017**
Also add two methods in message.py:  1. last_mle_message to find the newest mle; 2.assertAssignedRouterQuantity to confirm if Leader contains the Route64 TLV with 32 assigned router IDs.

### CODE REVIEW

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py [hunk 1]
```diff
@@ -31,61 +31,102 @@
...
+        for i in range(2, 32):
             self.nodes[i].start()
             time.sleep(5)
             self.assertEqual(self.nodes[i].get_state(), 'router')
 
-        self.nodes[DUT].start()
+        leader_messages = self.sniffer.get_messages_sent_by(DUT_LEADER)
+
+        # 2
+        self.nodes[32].start()
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: define `ROUTER_31 = 32 ` in advance and remove the magic number here?

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py [hunk 2]
```diff
@@ -31,61 +31,102 @@
...
+        time.sleep(5)
+        self.assertEqual(self.nodes[i].get_state(), 'router')
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: here not `i`

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py [hunk 3]
```diff
@@ -31,61 +31,102 @@
...
+
+        # 3 - DUT_LEADER
+        leader_messages = self.sniffer.get_messages_sent_by(DUT_LEADER)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: Duplicate with above Line 97.
Add comments here for easier understanding if it is on purpose?

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py [hunk 4]
```diff
@@ -31,61 +31,102 @@
...
+        msg.assertCoapMessageContainsTlv(network_layer.RouterMask)
+        msg.assertCoapMessageContainsTlv(network_layer.Rloc16)
+
+        status_tlv = msg.get_coap_message_tlv(network_layer.Status)
+        self.assertEqual(network_layer.StatusValues.SUCCESS, status_tlv.status)
+
+        # 4 - DUT_LEADER
+        msg = leader_messages.last_mle_message(mle.CommandType.ADVERTISEMENT)
+        msg.assertMleMessageContainsTlv(mle.Route64)
+        msg.assertAssignedRouterQuantity(32)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: `assertAssignedRouterQuantity()` would contain`assertMleMessageContainsTlv(mle.Route64)`, maybe Line #116 could be removed?

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py [hunk 5]
```diff
@@ -31,61 +31,102 @@
...
+
+        # 5 - Router_32
+        self.nodes[ROUTER_32].start()
         time.sleep(5)
-        self.assertEqual(self.nodes[DUT].get_state(), 'child')
+
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: why remove `child` state assertion?

File: tests/scripts/thread-cert/message.py [hunk 6]
```diff
@@ -166,6 +166,18 @@ def assertMleMessageContainsTlv(self, tlv_class_type):
                 break
 
         assert(contains_tlv == True)
+        return tlv
+
+    #To confirm if Leader contains the Route64 TLV with router_quantity assigned Router IDs.
```
Comment (librasungirl) [inline tests/scripts/thread-cert/message.py]: would it be better to update the comment as below?
> Verify the number of the active routers

File: tests/scripts/thread-cert/message.py [hunk 7]
```diff
@@ -294,8 +306,30 @@ def next_coap_message(self, code, uri_path=None, assert_enabled=True):
...
+            if m.type != MessageType.MLE:
+                continue
+
+            #for command_type in command_types:
+            if m.mle.command.type == command_type:
+                message = m
+                break
+
+        if assert_enabled:
+            assert message is not None, "Could not find CoapMessage with code: {}".format(command_type)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/message.py]: here `CoapMessage with code` should be `MleMessage with type`?

File: tests/scripts/thread-cert/message.py [hunk 8]
```diff
@@ -294,8 +306,30 @@ def next_coap_message(self, code, uri_path=None, assert_enabled=True):
 
         return message
 
+    #Get the newest mle message
```
Comment (librasungirl) [inline tests/scripts/thread-cert/message.py]: Would it be better to update the comment as below?
> Get the last Mle Message with specified type from existing capture.

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py [hunk 9]
```diff
@@ -31,61 +31,102 @@
...
+        leader_messages = self.sniffer.get_messages_sent_by(DUT_LEADER)
+
+        # 2
+        self.nodes[32].start()
+        time.sleep(5)
+        self.assertEqual(self.nodes[i].get_state(), 'router')
+
+        # 3 - DUT_LEADER
+        leader_messages = self.sniffer.get_messages_sent_by(DUT_LEADER)
+        msg = leader_messages.next_coap_message("2.04")
```
Comment (bukepo) [inline tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: prefer `'2.04'` here. single quoted.

File: tests/scripts/thread-cert/message.py (hunk 6 above)
Comment (bukepo) [inline tests/scripts/thread-cert/message.py]: This comment should go inside the function definition. same for `last_mle_message`

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 3 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: This method flushes the message queue so calling this method again will return only the newly logged messages.

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 2 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: great

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 5 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: Refer to Test Plan 5.2.3

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 1 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: fixed

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 2 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: fixed

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 3 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: fixed

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 4 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 6 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 8 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 7 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py (hunk 9 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_03_LeaderReject2Hops.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 6 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py [hunk 10]
```diff
@@ -294,8 +318,37 @@ def next_coap_message(self, code, uri_path=None, assert_enabled=True):
 
         return message
 
+    def last_mle_message(self, command_type, assert_enabled = True):
+        """Get the last Mle Message with specified type from existing capture.
+        Args:
```
Comment (librasungirl) [inline tests/scripts/thread-cert/message.py]: nit: leave a blank line before line #323 as the other places?

File: tests/scripts/thread-cert/message.py [hunk 11]
```diff
@@ -156,6 +156,14 @@ def assertMleMessageIsType(self, command_type):
         assert(self.mle.command.type == command_type)
 
     def assertMleMessageContainsTlv(self, tlv_class_type):
+        """To confirm if Mle message contains the  TLV.
```
Comment (librasungirl) [inline tests/scripts/thread-cert/message.py]: nit: remove the extra space before `TLV` ?
update `TLV` to `specified TLV type`?

File: tests/scripts/thread-cert/message.py [hunk 12]
```diff
@@ -166,6 +174,22 @@ def assertMleMessageContainsTlv(self, tlv_class_type):
                 break
 
         assert(contains_tlv == True)
+        return tlv
+
+    def assertAssignedRouterQuantity(self, router_quantity):
+        """ To confirm if Leader contains the Route64 TLV with router_quantity assigned Router IDs.
+
+        Args:
+            router_quantity: the quantity of router
```
Comment (librasungirl) [inline tests/scripts/thread-cert/message.py]: nit: add `. `in the end
and also other comments in this PR.

File: tests/scripts/thread-cert/message.py (hunk 11 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 12 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 10 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py [hunk 13]
```diff
@@ -156,6 +156,14 @@ def assertMleMessageIsType(self, command_type):
         assert(self.mle.command.type == command_type)
 
     def assertMleMessageContainsTlv(self, tlv_class_type):
+        """To confirm if Mle message contains the TLV type.
+
+        Args:
+            tlv_class_type: tlv's type.
+
+        Returns:
+            tlv: if contains the TLV, return it.
```
Comment (bukepo) [inline tests/scripts/thread-cert/message.py]: I guess `tlv` should be the type of the returned value.

File: tests/scripts/thread-cert/message.py [hunk 14]
```diff
@@ -294,8 +318,38 @@ def next_coap_message(self, code, uri_path=None, assert_enabled=True):
...
 
+    def last_mle_message(self, command_type, assert_enabled = True):
+        """Get the last Mle Message with specified type from existing capture.
+
+        Args:
+            command_type: the specified mle type.
+            assert_enabled: interrupt or not when get the mle.
+
+        Returns:
+            message: the last Mle Message with specified type.
```
Comment (bukepo) [inline tests/scripts/thread-cert/message.py]: `message` -> `Message`?

File: tests/scripts/thread-cert/message.py [hunk 15]
```diff
@@ -294,8 +318,38 @@ def next_coap_message(self, code, uri_path=None, assert_enabled=True):
...
+
+            #for command_type in command_types:
+            if m.mle.command.type == command_type:
+                message = m
+                break
+
+        if assert_enabled:
+            assert message is not None, "Could not find MleMessage with type: {}".format(command_type)
+
+
```
Comment (jwhui) [inline tests/scripts/thread-cert/message.py]: nit: remove extra line

File: tests/scripts/thread-cert/message.py [hunk 16]
```diff
@@ -166,6 +174,22 @@ def assertMleMessageContainsTlv(self, tlv_class_type):
                 break
 
         assert(contains_tlv == True)
+        return tlv
+
+    def assertAssignedRouterQuantity(self, router_quantity):
+        """ To confirm if Leader contains the Route64 TLV with router_quantity assigned Router IDs.
```
Comment (jwhui) [inline tests/scripts/thread-cert/message.py]: Does this follow [PEP 257](https://www.python.org/dev/peps/pep-0257/) docstring conventions?

File: tests/scripts/thread-cert/message.py (hunk 13 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 14 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

File: tests/scripts/thread-cert/message.py (hunk 16 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: I think I have fixed it. Could you please check it again?
I follow [url=https://google.github.io/styleguide/pyguide.html#Comments]Google Python Style Guide[/url] as before the code style.

Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: I think I have fixed it. Could you please check it again?
I follow [url=https://google.github.io/styleguide/pyguide.html#Comments]Google Python Style Guide[/url] as before the code style.[a](https://google.github.io/styleguide/pyguide.html)

File: tests/scripts/thread-cert/message.py (hunk 15 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/message.py]: fixed

**Pull request merged**